File: agency_management/agency_management/page/quotation_report/quotation_report.py
import frappe
from frappe.utils.pdf import get_pdf
import openpyxl
from openpyxl.styles import Alignment
import datetime


@frappe.whitelist()
def get_report_data(from_date, from_shift, to_date, to_shift):
    print("\n\n\n", from_date, from_shift, to_date, to_shift)
    print("\n\n\n\n\n\n\n\n\nMethod is called\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n")
    get_data = frappe.db.sql("""SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date='{}' AND shift IN('{}','Evening') AND docstatus=1
                            UNION SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date>'{}' AND  transaction_date<'{}' AND docstatus=1
                            UNION SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date='{}' AND shift IN('Morning','{}') AND docstatus=1
                        """.format(from_date, from_shift, from_date, to_date, to_date, to_shift), as_dict=1)
    # print(get_data,"\n\n\n\n\n\n")
    data = []
    data.extend(get_data)
    print("\n\n\n\n\n", data, "\n\n\n\n\n\n\n\n")
    context = {
        "records": data
    }
    template = frappe.render_template(
        "agency_management/public/js/page_templates/quotation_list.html", context)
#     print(template,"\n\n\n")
    return template


@frappe.whitelist()
def create_pdf(from_date, from_shift, to_date, to_shift):
    get_data = frappe.db.sql("""SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date='{}' AND shift IN('{}','Evening') AND docstatus=1
                            UNION SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date>'{}' AND  transaction_date<'{}' AND docstatus=1
                            UNION SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date='{}' AND shift IN('Morning','{}') AND docstatus=1
                        """.format(from_date, from_shift, from_date, to_date, to_date, to_shift), as_dict=1)
    data = []
    data.extend(get_data)
    context = {
        "records": data
    }
    template = frappe.render_template(
        "agency_management/public/js/page_templates/page_print_format.html", context)
    pdf = get_pdf(template)
    with open('/home/ruturaj/Downloads/report.pdf', 'wb') as file:
        file.write(pdf)
    return pdf


@frappe.whitelist()
def export_excel(from_date, from_shift, to_date, to_shift):
    # frappe.throw("Export to Excel Method is called")
    get_data = frappe.db.sql("""SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date='{}' AND shift IN('{}','Evening') AND docstatus=1
                            UNION SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date>'{}' AND  transaction_date<'{}' AND docstatus=1
                            UNION SELECT quotation_to, customer_name, transaction_date, shift, valid_till, contact_mobile, total, total_taxes_and_charges, grand_total FROM `tabQuotation`
                                    WHERE transaction_date='{}' AND shift IN('Morning','{}') AND docstatus=1
                        """.format(from_date, from_shift, from_date, to_date, to_date, to_shift), as_dict=1)
    datas = []
    datas.extend(get_data)
    print("\n\n\n\n\n\nData is fetched\n\n\n\n\n\n\n\n\n")
    wb = openpyxl.Workbook()
    print(wb)
    wb_sheet = wb.active
    print("\n\n\n\n\n\nSheet is  created\n",wb_sheet,"\n\n\n\n\n\n\n\n")

    GD=frappe.get_doc("Global Defaults")
    company_name=GD.default_company
    address=frappe.get_doc("Address",GD.default_company+"-billing")
    company_address=address.address_line1+", "+address.address_line2+", "+address.city+", "+address.state+", "+address.county
    
    # Add Sheet Head for company Information
    """ Here we Merge cell A1 to I2 and show company inforamtion"""
    wb_sheet.merge_cells('A1:I2')
    cell=wb_sheet.cell(row=1,column=1)
    cell.value = company_name+"\n"+company_address
    cell.alignment = Alignment(horizontal='center', vertical='center')



    # Add column headers
    wb_sheet["A3"]="Quotation To"
    wb_sheet["B3"]="Customer Name"
    wb_sheet["C3"]="Transaction Date"
    wb_sheet["D3"]="Shift"
    wb_sheet["E3"]="Valid Till"
    wb_sheet["F3"]="Contact"
    wb_sheet["G3"]="Total"
    wb_sheet["H3"]="Tax"
    wb_sheet["I3"]="Grand Total"

    # Style Column head
    # style_head=wb_sheet.cell(row=1,column=9)
    # style_head.font=style_head.font.copy(bold=True)


    # Add data in workbook
    row = 4
    for data in datas:
        wb_sheet["A"+str(row)]=data.get("quotation_to")
        wb_sheet["B"+str(row)]=data.get("customer_name")
        wb_sheet["C"+str(row)]=data.get("transaction_date").strftime("%d-%b-%Y")
        wb_sheet["D"+str(row)]=data.get("shift")
        wb_sheet["E"+str(row)]=data.get("valid_till").strftime("%d-%b-%Y")
        wb_sheet["F"+str(row)]=data.get("contact_mobile")
        wb_sheet["G"+str(row)]=data.get("total")
        wb_sheet["H"+str(row)]=data.get("total_taxes_and_charges")
        wb_sheet["I"+str(row)]=data.get("grand_total")
        row+=1

    # Save workbook
    wb.save("/home/ruturaj/Downloads/report.xlsx")

    return

File: agency_management/agency_management/page/quotation_report/quotation_report.js
frappe.pages['quotation-report'].on_page_load = function (wrapper) {
	new quotation_report(wrapper);
}

quotation_report = class QuotationReport {
	constructor(wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Quotation Report',
			single_column: true
		});
		this.parent = wrapper;
		this.page = this.parent.page;
		this.add_menu();
		this.add_filters();
	}
	add_menu() {
		const me = this;
		me.page.add_menu_item("PDF", ()=>this.create_pdf())
		me.page.add_menu_item("Export Excel", ()=>this.export_excel())
	}
	add_filters() {
		console.log("Make is called")
		const me = this;
		let start_date = me.page.add_field({
			label: "Start Date",
			fieldtype: "Date",
			fieldname: "start_date",
			reqd: 1,
			change() {
				me.start_date = start_date.get_value();
				me.make_report();
			}
		})
		let start_shift = me.page.add_field({
			label: "Start Shift",
			fieldtype: "Select",
			fieldname: "start_shift",
			reqd: 1,
			options: [
				'Morning',
				'Evening'
			],
			change() {
				me.start_shift = start_shift.get_value();
				me.make_report();
			}
		})
		let end_date = me.page.add_field({
			label: "End Date",
			fieldtype: "Date",
			fieldname: "end_date",
			reqd: 1,
			change() {
				me.end_date = end_date.get_value();
				me.make_report();
			}
		})
		let end_shift = me.page.add_field({
			label: "End Shift",
			fieldtype: "Select",
			fieldname: "end_shift",
			reqd: 1,
			options: [
				'Morning',
				'Evening'
			],
			change() {
				me.end_shift = end_shift.get_value();
				me.make_report();
			}
		})
	}

	make_report() {
		const me = this;
		// console.log(me.end_shift)
		if (me.start_date, me.start_shift, me.end_date, me.end_shift) {
			if (me.start_date < me.end_date) {
				console.log("All clear")
				frappe.call({
					method: "agency_management.agency_management.page.quotation_report.quotation_report.get_report_data",
					args: {
						"from_date": me.start_date,
						"from_shift": me.start_shift,
						"to_date": me.end_date,
						"to_shift": me.end_shift,
					},
					callback: function (r) {
						console.log(r.message)
						me.body = $(r.message).appendTo(me.page.main);
					}
				})
			}
			else {
				frappe.throw("Date are Invalid")
			}
		}
	}

	create_pdf(){
		const me = this;
		console.log("PDF called")
		if (me.start_date, me.start_shift, me.end_date, me.end_shift) {
			if (me.start_date < me.end_date) {
				frappe.call({
					method: "agency_management.agency_management.page.quotation_report.quotation_report.create_pdf",
					args: {
						"from_date": me.start_date,
						"from_shift": me.start_shift,
						"to_date": me.end_date,
						"to_shift": me.end_shift,
					}
				})
			}
			else {
				frappe.throw("Date are Invalid")
			}
		}
	}

	export_excel(){
		console.log("Export to excel called")
		const me = this;
		if (me.start_date, me.start_shift, me.end_date, me.end_shift) {
			if (me.start_date < me.end_date) {
				frappe.call({
					method: "agency_management.agency_management.page.quotation_report.quotation_report.export_excel",
					args: {
						"from_date": me.start_date,
						"from_shift": me.start_shift,
						"to_date": me.end_date,
						"to_shift": me.end_shift,
					}
				})
			}
			else {
				frappe.throw("Date are Invalid")
			}
		}
		frappe.msgprint("Excel is created")
	}


}